Java/cp/Crossword_puzzle.java

```diff
@@ -0,0 +1,244 @@
+            return false;
+        else if (i + word.length() < arr.length && arr[i + word.length()][j] != '+')
+            return false;
+
+        for (int ii = 0; ii < word.length(); ii++) {
+            if (i + ii >= arr.length)
+                return false;
+            if (arr[i + ii][j] == '-' || arr[i + ii][j] == word.charAt(ii))
+                continue;
+            else
+                return false;
+        }
+
+        return true;
+    }
+
+    // Places the word horizontally and returns a boolean array with only those
+    // indices marked true where changes were made by current word
+    public static boolean[] placeWordHorizontally(char[][] arr, String word, int i, int j) {
+        boolean[] wePlaced = new boolean[word.length()];
+        for (int jj = 0; jj < word.length(); jj++) {
+            if (arr[i][j + jj] == '-') {
+                arr[i][j + jj] = word.charAt(jj);
+                wePlaced[jj] = true;
+            }
+        }
+        return wePlaced;
+    }
+
+    // Places the word vertically and returns a boolean array with only those
+    // indices marked true where changes were made by current word
+    public static boolean[] placeWordVertically(char[][] arr, String word, int i, int j) {
+        boolean[] wePlaced = new boolean[word.length()];
+        for (int ii = 0; ii < word.length(); ii++) {
+            if (arr[i + ii][j] == '-') {
+                arr[i + ii][j] = word.charAt(ii);
+                wePlaced[ii] = true;
+            }
+        }
+        return wePlaced;
+    }
+
+    // Removing the current word from row
+    public static void removeWordHorizontally(char[][] arr, boolean[] wePlaced, int i, int j) {
+        for (int jj = 0; jj < wePlaced.length; jj++) {
+            if (wePlaced[jj] == true) {
+                arr[i][j + jj] = '-';
+            }
+        }
+    }
+
+    // Removing the current word from column
+    public static void removeWordVertically(char[][] arr, boolean[] wePlaced, int i, int j) {
+        for (int ii = 0; ii < wePlaced.length; ii++) {
+            if (wePlaced[ii] == true) {
+                arr[i + ii][j] = '-';
+            }
+        }
+    }
+}
+
+/*
+
+The approach behind this is to recursively check for each word in the vertical position and in the horizontal position. 
+Then fill the word in the matrix that can be the best fit in the corresponding position of the grid, 
+then update the crossword grid by filling the gap with that word.
+
+Sample Input/Output
+
+Example 1: 
+
+Enter the crossword puzzle:
++-++++++++
++-++++++++
++-++++++++
++-----++++
++-+++-++++
++-+++-++++
++++++-++++
+++------++
++++++-++++
++++++-++++
+
+Enter the total words:
+4
+
+Enter the words:
+LONDON
+DELHI
+ICELAND
+ANKARA
+
+1.  + L + + + + + + + +
+    + O + + + + + + + +
+    + N + + + + + + + +
+    + D E L H I + + + +
+    + O + + + C + + + +
+    + N + + + E + + + +
+    + + + + + L + + + +
+    + + A N K A R A + +
+    + + + + + N + + + +
+    + + + + + D + + + +
+
```

Comment:
Time & Space Complexity are missing. Please write it down at then end in the commented form

Reply:
Done, please check!